### DISCUSSION TITLE
buffering

### DISCUSSION THREAD

**Discussion by zdzisius** [schreibfaul1/ESP32-audioI2S#390] [Q&A]
Hi,

I am using this wonderful library for my radio that i built. Very good job and thanks for sharing it. Too bad that aac+ format is not supported.
Anyway, I have a question. Does anybody know how to set buffering to 1 or 2 seconds for radio connection? This way I would like to resolve the issue with drop outs for poor connection. Maybe someone has already solved this problem?

Thanks in advance

**Comment by Kiwi915** [comment]
Hi, I also have drop out problems.
Do you have a esp with PSRam  or a normal Esp32 ?

**Reply by zdzisius** [reply]
I got esp32-wroom and I also gonna have esp32-wrover, because I've already know that PSRam is needed for buffering.. But I am not sure if default settings in audio.h library is enought for poor ineternet connection.. that's why I am considering to expand buffor to at least 2s.

**Comment by Kiwi915** [comment]
Hi, many thanks for the answer.Can you please advise me were I have to expand the 2s ?
Can you show me the line in code please?
Kind regards.....

**Reply by zdzisius** [reply]
I am not sure for now where in code  you can change the buffor, but you need to have esp with PSram first (like wrover). 
My blind guessing is that you can try to change values around line 140 in the code of audio.h. But I am not so experienced in it yet, so it would be good if somebody who has more experience in this topic could confirm what I've already wrote here.

**Comment by Kiwi915** [comment]
Ok I understand. Iam going to buy a Wroover and test again with the PSRam. Hopefully anybody knows in the community how to expand the time buffer in the code.
Kind regards and thanks to you again...

**Reply by Kiwi915** [reply]
Hi, I want to update all.
With the Wroover I have nearly zero slow streams.
Regards

**Reply by zdzisius** [reply]
@Kiwi915 Ok, but you are using the default settings and didn't change the buffor at all?

**Comment by schreibfaul1** [comment]
One can compare the function of the audio buffer with the leaky bucket model. The amount of water that drains from the bucket is predetermined. For an mp3 stream, this is the compression rate and the bit rate. And the mechanism that fills the bucket is the WiFClient in our case. And there is the bottleneck. If the bottom of the bucket becomes visible, in the mp3 example there is less audio data than is likely to fit in an mp3 frame, a warning is issued. The size of the bucket is not decisive, sufficient water must flow.
There are discussions about how the WiFiClient's performance can be increased, e.g. through asynchronous access or by changing the buffer size via sdkconfig.
The WifiClient is part of the ESP Arduino core and is easy to use. And the changes must also take effect in WiFiClientSecure.
